src/ui/menu/credits/credits_popup.cpp

```diff
@@ -39,87 +42,144 @@ bool GlobedCreditsPopup::setup() {
+                    credit["name"].as_string(),
+                    credit["accountID"].as_int(),
+                    credit["userID"].as_int(),
+                    credit["iconID"].as_int(),
+                    credit["color1"].as_int(),
+                    credit["color2"].as_int(),
+                    credit["color3"].as_int());
+            }
+
+            auto* cellStaff = Build<GlobedCreditsCell>::create("Staff", false, staff)
+                .id("cell-staff"_spr)
+                .parent(scrollLayer->m_contentLayer)
+                .collect();
+
+            /* Contributor */
+
+            CCArray* contributors = CCArray::create();
+
+            auto contrib_credits = credits_data["contrib"].as_array();
+
+            for (auto& credit : contrib_credits) {
+                ADD_PLAYER_NICK(
+                    contributors,
+                    credit["name"].as_string(),
+                    credit["name"].as_string(),
+                    credit["accountID"].as_int(),
+                    credit["userID"].as_int(),
+                    credit["iconID"].as_int(),
+                    credit["color1"].as_int(),
+                    credit["color2"].as_int(),
+                    credit["color3"].as_int());
+            }
+
+            auto* cellContrib = Build<GlobedCreditsCell>::create("Contributor", true, contributors)
+                .id("cell-contributor"_spr)
+                .parent(scrollLayer->m_contentLayer)
+                .collect();
+
+            /* Special */
+
+            CCArray* special = CCArray::create();
+
+            auto special_credits = credits_data["special"].as_array();
+
+            for (auto& credit : special_credits) {
+                ADD_PLAYER_NICK(
+                    special,
+                    credit["name"].as_string(),
+                    credit["name"].as_string(),
+                    credit["accountID"].as_int(),
+                    credit["userID"].as_int(),
+                    credit["iconID"].as_int(),
+                    credit["color1"].as_int(),
+                    credit["color2"].as_int(),
+                    credit["color3"].as_int());
+            }
+
+            auto* cellSpecial = Build<GlobedCreditsCell>::create("Special thanks", false, special)
+                .id("cell-special"_spr)
+                .parent(scrollLayer->m_contentLayer)
+                .collect();
+
+            scrollLayer->m_contentLayer->setContentHeight(
+                cellOwners->getScaledContentSize().height
+                + cellDevs->getScaledContentSize().height
+                + cellStaff->getScaledContentSize().height
+                + cellContrib->getScaledContentSize().height
+                + cellSpecial->getScaledContentSize().height
+            );
+            scrollLayer->m_contentLayer->updateLayout();
+
+            // make the list start at the top instead of the bottom
+            util::ui::scrollToTop(scrollLayer);
```

Comment:
tbh i think it'd be cleaner to just move this to a function in CreditsPopup and just call it, instead of having a massive web request callback

Reply:
also, unexpected json data will 100% crash the game. it would probably be a super rare edge case, but i would still prefer that it's validated

the best solution is probably making a separate struct for credit entries and implement matjson deserializer for it. would make the code a lot cleaner.

Reply:
side note: why is the "name" key read twice? please note that display name and username might not always match, the display name is what's shown and username is what's passed to `GameLevelManager::storeUserName`, so that if you open their profile, it doens't show a '-' and shows their actual GD name.